lib/core/util.js

```diff
@@ -436,6 +436,42 @@ function parseHeaders (headers, obj) {
   return obj
 }
 
+/**
+ * @param {Record<string, string | string[]>} headers
+ * @returns {Buffer[]}
+ */
+function encodeHeaders (headers) {
+  const headerNames = Object.keys(headers)
+
+  /**
+   * @type {Buffer[]}
+   */
+  const rawHeaders = new Array(headerNames.length * 2)
+
+  let rawHeadersIndex = 0
+  for (const header of headerNames) {
+    let rawValue
+    const value = headers[header]
+    if (Array.isArray(value)) {
+      rawValue = new Array(value.length)
+
+      for (let i = 0; i < value.length; i++) {
+        rawValue[i] = Buffer.from(value[i])
+      }
+    } else {
+      rawValue = Buffer.from(value)
+    }
+
+    rawHeaders[rawHeadersIndex] = Buffer.from(header)
```

Comment:
we have in lib/core/constants wellknownheadernames. Maybe we should create buffers from them and reuse them if possible to avoid creating buffers over and over.